### PR TITLE
Example: spectral clustering based graph

### PR DESCRIPTION
**Pull request by sharwinbobde**
**[Updated]**
This example is meant to address Issue #4261
More specifically, using scikit-learn to cluster data using Spectral Clustering. Then using the adjacency matrix from the same to generate a graph for the data points to understand the clustering.

image

Follow the below thread and the issue to view the iterations of the demo.

### CODE REVIEW

**Comment by sharwinbobde** [comment]
@dschult we changed the colour schemes for plots of Step 1 and 2 to indicate that because clustering is an unsupervised method, the grouping does not necessarily indicate class membership.

**Comment by dschult** [comment]
This is much clearer--thank you.

But it still is not clear how you are assigning the edges in a network.  It looks like you start with a dataset with three variables for each observation ==> two are continuous variables and the third is a binary classification. Then you use an AI method (which one?) to classify the observations into two groups. The clustering output shows these groups and they are close but not the same as the original groups. Is this correct?

Then you show the network -- but there is no explanation for what the nodes represent, how the edges are chosen, etc.

**Comment by sharwinbobde** [comment]
I'll write a detailed flow here first thing in the morning :)

**Comment by sharwinbobde** [comment]
So here is what we decided to do
## Task overview
- We do everything on a toy dataset
- We know data comes from 2 distributions only to compare results.
- The data is noisy (to show actual problems with clustering).
- We will use Spectral Clustering to (I) understand the data and (ii) cluster it
- We can also build a network using the affinity matrix which results in the following:
     - nodes are data points from the feature space
     - edges' weight is the value in the affinity matrix
     - nodes have cluster membership information
     
#### Usage
- This can be used to understand unlabelled data in a network form. For example we can represent a 10-dimensional dataset using networks while preserving local information and making it obvious.
- The resulting graph is a condensed form of the feature space information
- We can indirectly use graph algorithms on feature space information
- Edge weights do not have small noisy information from feature measurements, only the neighbourhood relation information.
 

## Step 1: Generating sample dataset
- Generate 2-class data which is not linearly separable 
```
n_samples = 100
X, y = datasets.make_moons(n_samples=n_samples)
```
- Add gaussian noise to make the problem difficult: `X = X + np.random.normal(1, 0.1, X.shape)`
- The problem can look much cleaner with really clean clustering if we reduce the noise, but the readers will need to know that clustering has its limitations and not to use it blindly.

## Step 2: Clustering
- Lets conveniently select **2 as the number of clusters**. (this is just an example/demonstration :) )
- We will use *nearest-neighbour* based affinity matrix for [SpectralClustering](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.SpectralClustering.html). This is what the method does under the hood:
     - Given only the feature space information, find 10 nearest neighbours for each data point using euclidean distance.
     - create the affinity matrix. Here is the code snippet from the scikit-learn source code. 
     ```
     connectivity = kneighbors_graph(X, n_neighbors=self.n_neighbors,
                                            include_self=True,
                                            n_jobs=self.n_jobs)
     self.affinity_matrix_ = 0.5 * (connectivity + connectivity.T)
     ```

     - this is what a sample affinity sparse matrix looks like:  
```
    (0, 76) 0.5
    (0, 70) 0.5
    (0, 61) 0.5
    (0, 36) 0.5
    (0, 66) 1.0
    (0, 92) 1.0
    (0, 15) 1.0
    ...
    (98, 83) 1.0
    (98, 99) 1.0
    (98, 98) 1.0
    (99, 25) 1.5
    (99, 20) 1.0
    (99, 48) 1.0
```
- label the data points according to the cluster membership
     
   
## Step 3: constructing the network
- get the affinity matrix
- get the labels
- remove self-loops
```
    clusters = sc.fit(X)
    cluster_affinity_matrix = clusters.affinity_matrix_
    print(cluster_affinity_matrix)

    pred_labels = clusters.labels_.astype(int)
    G = nx.from_scipy_sparse_matrix(cluster_affinity_matrix)
    # remove self edges
    G.remove_edges_from(nx.selfloop_edges(G))
```

*Then just plot the different steps*

**Comment by sharwinbobde** [comment]
_Oops! posted the messages on the issue thread instead of this PR before.
Deleted the content on the Issue to keep things clean._

@dschult hope this is clearer :)
If there are any doubts we will keep updating the above explanation.
We do believe from our side that this is a strong direction for the Spectral Clustering example because it covers many possibilities of usage while still showing the limitations.

**Comment by dschult** [comment]
Much more complete information here...  Thank you!

The original description had the sentence: "Then using the data from the spectral clustering to generate a graph for the data points."   If I understand correctly, the "graph" is constructed using the affinity matrix by computing k-nearest-neighbors.  But maybe here you mean "graph" (as in draw) a scatter plot of the feature space showing the clusters. In general, stay away from the word "graph" to mean a scatter plot when you've got graph theory in the mix. Graph Theory graphs can be called "networks", while graphs of datasets can be called scatter plots (or whatever other plot they are).

What is the right visualization of the network that will make this algorithm more clear?  I think you have formed the network using distance in the feature space between data points. When you plot it you use a spring_layout for the nodes (which are the data points).  Maybe it would be better to use the feature values as the coordinates of the nodes.  Then you will show the essentially the scatter plot with edges between neighbors. 

Another approach would be to plot the nodes using laplacian eigenvector values for two main eigenvalues. That gives a spectral layout which may (or may not depending on the meaning of "spectral clustering") show how the clustering is being determined. 

Perhaps both pictures of the network would be helpful in getting the reader to understand 1) where the matrix/network is coming from, and 2) where the clustering is coming from once the network is formed.

What do you think?  Or maybe you can explain why you chose to show the axes and positions you chose -- what is your pedagogical goal and how are you working to attain it?

**Review by rossbar — COMMENTED** [review]
> representing items in a network and creating avenues to using graph algorithms on feature-space information.

This example could be made much more compelling by actually applying a graph algorithm on feature-space information. The one high-level comment I have is that it's not clear why networkx is even necessary, since the data generation and analysis (clustering) are both done w/ scikit-learn. It's nice to be able to explicitly see the edges after converting to a networkx graph, and the demonstration of the layout is also nice, but in its current form this seems more like a pure scikit-learn example than a scikit-learn + networkx example.

That's not to say it isn't valuable though! I'm not an expert so it's quite possible I'm missing interesting details - in that case adding some narrative text in the form of rst comments might help.

**Comment by sharwinbobde** [comment]
I see that some choices we made can be a bit confusing and deliver the wrong message. We will try to address these with a modified example.
image

Here you can clearly see that.
- Step 1. Some data exists. we know there are 3 clusters.
- Step 2. We do Spectral clustering (based on the nearest neighbours method) with the number of clusters as 3.
- Step 3. We plot the graph obtained from the affinity matrix.

Modification considerations:
- clustering is more apparent with consistent colour schemes.
- it is more apparent that edges have neighbour information
- what is a scatterplot and what is a simple network is clearer.
- use case is comparatively more apparent.


This example can be used to observe cluster quality for high-dimensional data using a 2d network plot.
@rossbar I agree with you that there should be a Step 4. where we use some sort of graph technique.

Clarifications for comments:
- @dschult This modification makes the distinction between scatterplots and network plots clearer.
- @dschult @rossbar the affinity matrix will bot store real distance information at the clustering phase.
     - edges exist only if two nodes are neighbours
     - the edge weight is either **0.5** (if one node  is neighbour of the other) or **1.0** (if both nodes are neighbours of each other)
     - Thus, the network representation has information that depends on feature-space distances but is NOT equal to the euclidean distance. _The graph is a condensed form of information_.
- @rossbar In Step 3. We represent neighbours for different cluster members and thus draw the condensed high dimensional feature information using networkx, that is separated properly.
- @rossbar for step 4. we might compute some heuristics to show clustering quallity.

**Comment by sharwinbobde** [comment]
I'll update the above "Demonstration Explanation"

**Comment by sharwinbobde** [comment]
image

One more Iteration!
In Step 4 we record the cut-size for different cluster-pair and plot the confusion matrix. The above figure shows that the cut size of clusters 2 and 1 is 0 so they are cleanly separated, whereas clusters 2 and 0 have a lot of similar neighbours. This cut-size is the heuristic Data Scientist can use to actually see how clean the clustering algorithm's output is and quantifiably measure the how good the clustering is.

**Comment by rossbar** [comment]
One thing to keep in mind is that the example should highlight something about graph/network analysis. There is a fine line to walk between making an example compelling and making it understandable. For me anyways, making the example *more* complex is heading in the wrong direction :). That's not to say that it isn't valuable - in fact, as the scope of the example here continues to increase, it may be a better fit for the [networkx-notebooks](https://github.com/networkx/notebooks) repo, which is a more natural fit for larger examples and examples that would benefit from some narrative/explanatory text.

Also - there's no need to continue attaching images to the GH comments - all examples are run by the CI and can be previewed by clicking on the "ci/circleci: build artifact" tab in the CI check summary. This is actually better for review because it also guarantees that the image is being generated as expected (which is [not currently the case](https://1592-890377-gh.circle-artifacts.com/0/doc/build/html/auto_examples/index.html#external-libraries)).

**Comment by lobpcg** [comment]
The PR appears abandoned, and the code no longer runs on the current release.

File: examples/external/spectral_clustering_graph.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+import networkx as nx
+import seaborn as sns
+
+np.random.seed(0)
+
+
+def spectral_clustering_to_graph(X, n_clusters: int, sc: SpectralClustering):
+    # get affinity matrix from spectral clustering
+    clusters = sc.fit(X)
+    cluster_affinity_matrix = clusters.affinity_matrix_
```
Comment (lobpcg) [inline examples/external/spectral_clustering_graph.py:24]: cluster_affinity_matrix = clusters.affinity_matrix_.H

File: examples/external/spectral_clustering_graph.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    pred_labels = clusters.labels_.astype(int)
+    G = nx.from_scipy_sparse_matrix(cluster_affinity_matrix)
```
Comment (lobpcg) [inline examples/external/spectral_clustering_graph.py:27]: G = nx.from_scipy_sparse_array(cluster_affinity_matrix)

**Comment by rossbar** [comment]
Thanks for the ping @lobpcg - I still have the sense that this is a better fit for a scikit-learn gallery rather than networkx. Since it's been 18 months since last activity and is still in draft, I'm inclined to close. I'll do so soon if there are no objections!

**Comment by lobpcg** [comment]
@rossbar Yes, I agree that this PR could be closed as stalled. I like the idea, but the proposed implementation is still far from being acceptable unfortunately.

**Comment by lobpcg** [comment]
@rossbar I have made https://github.com/networkx/networkx/pull/7040 replacing this PR - could you please have a look?
